DOCSP-49230 - Aggregation stages (builders) (#591)
mongodb/docs-csharp: source/aggregation/stages.txt

.. _csharp-aggregation-stages:

==================
Aggregation Stages
==================

.. facet::
   :name: genre
   :values: reference

.. meta::
   :keywords: dotnet, code example, transform, pipeline

.. contents:: On this page
   :local:
   :backlinks: none
   :depth: 2
   :class: singlecol

Overview
--------

> On this page, you can learn how to create an aggregation pipeline and add stages to it
> by using methods in the {+driver-short+}.

Reviewer: S: since this guide is more about creating the stages than adding them to a pipeline
```suggestion
On this page, you can learn how to create an aggregation pipeline and pipeline stages
by using methods in the {+driver-short+}.
```

You can build a type-safe aggregation pipeline in the following ways:

- Construct an ``EmptyPipelineDefinition<TDocument>`` object. Chain calls from this object
  to the relevant aggregation methods. Then, pass the pipeline object to the
  ``IMongoCollection<TDocument>.Aggregate()`` method.

- Call the ``IMongoCollection<TDocument>.Aggregate()`` method. Chain calls from this
  method call to the relevant aggregation methods.

Select the :guilabel:`EmptyPipelineDefinition`
or :guilabel:`Aggregate` tab to see the corresponding code for each approach:

.. tabs::

   .. tab:: EmptyPipelineDefinition
      :tabid: empty-pipeline-definition

      .. code-block:: csharp

         // Defines the aggregation pipeline
         var pipeline = new EmptyPipelineDefinition<Movie>()
            .Match(...)
            .Group(...)
            .Merge(...);

         // Executes the aggregation pipeline
         var results = collection.Aggregate(pipeline).ToList();

   .. tab:: Aggregate
      :tabid: aggregate

      .. code-block:: csharp

         // Defines and executes the aggregation pipeline
         var results = collection.Aggregate()
            .Match(...)
            .Group(...)

>             .Merge(...);

Reviewer: S: should you also call `ToList()` at the end here?

Author: It's probably better to remove the call from the other example.

.. _csharp-aggregation-stages-bsondocument:

BsonDocument
~~~~~~~~~~~~

Some aggregation stages don't have corresponding methods in the {+driver-short+}.
To add these stages to your pipeline, use ``BsonDocument`` objects or string literals
to construct a stage in the Query API syntax. Then, pass the BSON document to the
``PipelineDefinitionBuilder.AppendStage()`` method. This syntax supports all stages
in the aggregation pipeline, but doesn't provide type hints or type safety.

The following code example shows how to add ``$unset``, an aggregation

> stage without a corresponding method, to an empty aggregation pipeline:

Reviewer: ```suggestion
stage without a corresponding builder method, to an empty aggregation pipeline:
```

Author: LMK if you think more is needed

.. code-block:: csharp

>    var pipeline = new EmptyPipelineDefinition<BsonDocument>().AppendStage<BsonDocument, BsonDocument, BsonDocument>("{ $unset: "field1" }");

Reviewer: S: consider newlining the method calls

> .. note::

Reviewer: S: consider changing to important

   If you use a ``BsonDocument`` to define a pipeline stage, the driver doesn't
   recognize any ``BsonClassMap`` attributes, serialization attributes, or
   serialization conventions. The field names that you use in the ``BsonDocument`` must
   match the field names stored in {+mdb-server+}.

Aggregation Stage Methods
-------------------------

> The following table lists the builders methods in the {+driver-short+} that correspond

Reviewer: ```suggestion
The following table lists the builder methods in the {+driver-short+} that correspond
```

> to stages in the aggregation pipeline. For more information about an aggregation stage,
> click the stage name. For more information about a builders method, click the
> method name. If an aggregation stage isn't in the table, you must use the
> :ref:`BsonDocument <csharp-aggregation-stages-bsondocument>` syntax to add the stage
> to your pipeline.

Reviewer: ```suggestion
to stages in the aggregation pipeline. To learn more about an aggregation stage,
follow the link from the stage name to the Server manual reference. To learn more about a
builder method, follow the link from the method name to the dedicated guide. If an
aggregation stage isn't in the table, the driver does not provide a builder method, so you must use the
:ref:`BsonDocument <csharp-aggregation-stages-bsondocument>` syntax to add the stage
to your pipeline.
```

to your pipeline.

.. list-table::
   :header-rows: 1
   :widths: 20 80

  * - Aggregation Stage
    - Description
    - Builders Method

  * - :manual:`$bucket </reference/operator/aggregation/bucket/>` 

>     - Categorizes incoming documents into groups, called buckets,

Author: Stage descriptions were taken (almost) verbatim from server docs

      based on a specified expression and bucket boundaries.

>     - :ref:`Bucket() <csharp-aggregation-bucket>`

Author: All method links will be broken for the time being

     - :ref:`ChangeStreamSplitLargeEvent() <csharp-aggregation-changestreamsplitlargeevent>`

   * - :manual:`$count </reference/operator/aggregation/count/>`
     - Returns a count of the number of documents at this stage of
       the aggregation pipeline.
     - :ref:`Count() <csharp-aggregation-count>`

   * - :manual:`$densify </reference/operator/aggregation/densify/>`
     - Creates new documents in a sequence of documents where certain values in a field are missing.
     - :ref:`Densify() <csharp-aggregation-densify>`

   * - :manual:`$documents </reference/operator/aggregation/documents/>`
     - Returns literal documents from input expressions.
     - :ref:`Documents() <csharp-aggregation-documents>`

   * - :manual:`$facet </reference/operator/aggregation/facet/>`
     - Processes multiple aggregation pipelines
       within a single stage on the same set
       of input documents. Enables the creation of multi-faceted
       aggregations capable of characterizing data across multiple
       dimensions, or facets, in a single stage.
     - :ref:`Facet() <csharp-aggregation-facet>`

   * - :manual:`$graphLookup </reference/operator/aggregation/graphLookup/>`

>      - Performs a recursive search on a collection. To each output
>        document, adds a new array field that contains the traversal
>        results of the recursive search for that document.

Reviewer: ```suggestion
     - Performs a recursive search on a collection. This method adds
       a new array field to each output document that contains the traversal
       results of the recursive search for that document.
```

     - Randomly selects the specified number of documents from its
       input.
     - :ref:`Sample() <csharp-aggregation-sample>`

   * - :manual:`$search </reference/operator/aggregation/search/>`
     - Performs a full-text search of the field or fields in an
       :atlas:`Atlas </reference/atlas-search/query-syntax/>`
       collection.

       This stage is available only for MongoDB Atlas clusters, and is not
       available for self-managed deployments. To learn more, see
       :atlas:`Atlas Search Aggregation Pipeline Stages
       </reference/atlas-search/query-syntax>` in the Atlas documentation.
     - :ref:`Search() <csharp-aggregation-search>`

   * - :manual:`$searchMeta </reference/operator/aggregation/searchMeta/>`
     - Returns different types of metadata result documents for the
       :atlas:`Atlas Search </atlas-search>` query against an
       :atlas:`Atlas </reference/atlas-search/query-syntax/>`
       collection.

       This stage is available only for MongoDB Atlas clusters,
       and is not available for self-managed deployments. To learn
       more, see :atlas:`Atlas Search Aggregation Pipeline Stages

>        </reference/atlas-search/query-syntax>` in the Atlas documentation.

Reviewer: S: add this info to the vectorSearch stage entry at the end of this table

     - :ref:`SortByCount() <csharp-aggregation-sortbycount>`

   * - :manual:`$unionWith </reference/operator/aggregation/unionWith/>`
     - Combines pipeline results from two collections into a single
       result set.
     - :ref:`UnionWith() <csharp-aggregation-unionwith>`

   * - :manual:`$unwind </reference/operator/aggregation/unwind/>`
     - Deconstructs an array field from the input documents to
       output a document for *each* element. Each output document
       replaces the array with an element value. For each input
       document, outputs *n* Documents, where *n* is the number of
       array elements. *n* can be zero for an empty array.
     - :ref:`Unwind() <csharp-aggregation-unwind>`

   * - :manual:`$vectorSearch </reference/operator/aggregation/vectorSearch/>`
     - Performs an :abbr:`ANN (Approximate Nearest Neighbor)` or
       :abbr:`ENN (Exact Nearest Neighbor)` search on a
       vector in the specified field of an
       :atlas:`Atlas </reference/atlas-search/query-syntax/>` collection.
     - :ref:`VectorSearch() <csharp-aggregation-vectorsearch>`

API Documentation
-----------------

To learn more about assembling an aggregation pipeline, see

> :manual:`Aggregation Pipeline </core/aggregation-pipeline/>`.

Reviewer: S: add link target
```suggestion
:manual:`Aggregation Pipeline </core/aggregation-pipeline/>` in the Server manual.
```

Reviewer: applies everywhere

.. tabs::

   .. tab:: EmptyPipelineDefinition
      :tabid: empty-pipeline-definition

      .. code-block:: csharp

         // Defines the aggregation pipeline
         var pipeline = new EmptyPipelineDefinition<Movie>()
            .Match(...)
            .Group(...)
            .Merge(...);

         // Executes the aggregation pipeline
         var results = collection.Aggregate(pipeline);

   .. tab:: Aggregate
      :tabid: aggregate

      .. code-block:: csharp

         // Defines and executes the aggregation pipeline
         var results = collection.Aggregate()
            .Match(...)
            .Group(...)
            .Merge(...);

.. _csharp-aggregation-stages-bsondocument:

BsonDocument
~~~~~~~~~~~~

Some aggregation stages don't have corresponding methods in the {+driver-short+}.
To add these stages to your pipeline, use ``BsonDocument`` objects or string literals
to construct a stage in the Query API syntax. Then, pass the BSON document to the
``PipelineDefinitionBuilder.AppendStage()`` method. This syntax supports all stages
in the aggregation pipeline, but doesn't provide type hints or type safety.

> The following code example shows how to add ``$unset``, an aggregation
> stage without a corresponding builder method, to an empty aggregation pipeline:

Reviewer: S; wording still awkward, maybe try this wording as its actually not super important to know that the stage doesnt have a builder method (since you can do this with any stage)
```suggestion
The following code example shows how to add an ``$unset`` stage to an empty aggregation pipeline:
```

.. code-block:: csharp

   var pipeline = new EmptyPipelineDefinition<BsonDocument>()

>            .AppendStage<BsonDocument, BsonDocument, BsonDocument>("{ $unset: "field1" }");

Reviewer: S: text highlighting weird, do you need to escape the characters as there's multiple sets of " marks?

   * - :manual:`$unionWith </reference/operator/aggregation/unionWith/>`
     - Combines pipeline results from two collections into a single
       result set.
     - :ref:`UnionWith() <csharp-aggregation-unionwith>`

   * - :manual:`$unwind </reference/operator/aggregation/unwind/>`
     - Deconstructs an array field from the input documents to
       output a document for *each* element. Each output document
       replaces the array with an element value. For each input
       document, outputs *n* Documents, where *n* is the number of
       array elements. *n* can be zero for an empty array.
     - :ref:`Unwind() <csharp-aggregation-unwind>`

   * - :manual:`$vectorSearch </reference/operator/aggregation/vectorSearch/>`
     - Performs an :abbr:`ANN (Approximate Nearest Neighbor)` or
       :abbr:`ENN (Exact Nearest Neighbor)` search on a
       vector in the specified field of an
       :atlas:`Atlas </reference/atlas-search/query-syntax/>` collection.

       This stage is available only for MongoDB Atlas clusters, and is not
       available for self-managed deployments. To learn more, see
       :atlas:`Atlas Search Aggregation Pipeline Stages

>        </reference/atlas-search/query-syntax>` in the Atlas documentation.

Reviewer: S: change the link to go to the atlas vector search overview
